### ISSUE TITLE
Order of execution (injector & plugins)

### ISSUE THREAD

**Issue by pwnosaur** [tklengyel/drakvuf#1158]
Hi,

During some tests, I have been using the LibInjector directly instead of using it through DRAKVUF switches, the reason behind this was the order of execution, I noticed that DRAKVUF starts the injector before starting the plugins which results in missing some system calls, for example it misses the call to `NtCreateUserProcess` that was initiated from the targeted injection process.

I can't see the reason why it was implemented in that order; if it wouldn't break anything, would it be Ok to make a pull request and switch the order ? as I believe it would be better not to miss anything the plugins want to capture.

A practical reason for that suggestion would be interception of the injection and based on receiving the process creation signal, the plugin follow certain execution path or influence the creation of the process itself.

**Comment by icedevml** [comment]
I think `injector` was too unstable back when it was decided to be implemented this way and also there were few bugs around altp2m handling that could prevent that from working fine.

`injector` is pretty invasive, it hooks CR3 callback and some system calls, diverting the flow of the hijacked program. I think it might work if you would reorder but YMMV. Also, injector generally needs a big rewrite(tm).

**Comment by pwnosaur** [comment]
> `injector` is pretty invasive, it hooks CR3 callback and some system calls, diverting the flow of the hijacked program. I think it might work if you would reorder but YMMV. Some more invasive plugins might disrupt injector for some reason. Also, injector generally needs a big rewrite(tm).

Well in that case it may be possible is to allow the plugins to have configuration that set it to either start before or after injection, and by default it would be after to ensure it won't break anything on the current version, and for those plugins that want to start before the injection, it can be simply modified in the configuration.

**Comment by icedevml** [comment]
I bet `syscalls` would work, more invasive ones like `apimon` - really not sure.

If the `syscalls` would not work then it would be some bug in the engine or injector logic, but I could help debugging this if you would have some PoC branch.

**Comment by tklengyel** [comment]
Generally speaking anything building on top of libdrakvuf will "just work", they don't have to be aware of each other. Of course that stands for passive monitoring that don't mess with the VM state (ie write to the memory, change registers). Once you start messing with the VM all bets are off. So that's why injector comes before the plugins. Plugins are (mostly) passive monitors. Injector is very invasive.

Also, the reason the plugins start after the injector is that the injected process doesn't start execution right away. It's injected with the `CREATE_SUSPENDED` flag. When we see the process injection worked and the process is ready to go we inject the resume call and also start the plugins. We only unpause the VM when all plugins are good to go. That way you won't ever miss any of the process' actual execution, only the creation bits that were done by loader.

**Comment by icedevml** [comment]
Yup, and while the creation bits might be not trivial, I think they are mostly predictable, because injector implementation is known and the hijacked process (e.g. explorer, task manager) is usually well-known as well. Unless you don't need to know very low level details about the injection itself, I think there would be no huge benefit from this.

**Comment by pwnosaur** [comment]
Well I have added the support for such capability to choose to start the plugins before the injection using '-z' as a switch, it may not be necessary after what @tklengyel  explained, but giving the option to the user might not be a bad idea after all, the modifications I have made shouldn't break anything or change it from its current state.

https://github.com/pwnosaur/drakvuf/tree/ExecutionOrder

**Comment by tklengyel** [comment]
Thanks but I think this would just needlessly complicate startup right now. Without a clear need to start a plugin before injection happens I think for now we should keep things as-is. Once there is a compelling reason to have this option I would be happy to reconsider.

**Comment by icedevml** [comment]
I don't have strong opinion here but I'm for avoiding complexity as well.
